### PR TITLE
Bug fix of concatenate + max

### PR DESCRIPTION
**Pull request by yukoba**
I think this pull request will solve https://github.com/HIPS/autograd/issues/138 .
```argmax_locations```should not include same value indices.
Only first one should be included.

But this might be slow for a large array because I'm using ```O(len(idx) ** 2)``` for loops.

If this pull request seems correct, I will modify https://github.com/HIPS/autograd/pull/140 .

### CODE REVIEW

**Comment by yukoba** [comment]
I modified the algorithm from O(len(idx) ** 2) to O(len(idx)). Become faster.

**Comment by duvenaud** [comment]
Yukoba, you're the best.

As far as I understand, this is a real bug that occurs when we take the max over an array that can have duplicates of the same variable.  So a simpler function that would have had the same problem is:

```
def f(x):
    return np.max(np.stack([x x]))
```

The same problem also exists for np.maximum(x, x), although I don't know if that's worth fixing.

Dougal and Matt: Even though this patch'll make max() slower, since the bug came up in a real use case, I vote for merge.

**Comment by mattjj** [comment]
Was the only reason to write it that way originally to make it fast?

Just depending on the first does make sense according to numpy's semantics (according to the docstring of `np.argmax`), though I think arguments could be made for following other conventions. The root of the problem is that if we think of `np.max` as a function from R^2 to R, with input (x,y) and output z, then it's not smooth along y=x and so the 'right' gradient is ambiguous. One reasonable thing to do is to output an element of the subgradient, which is the convex hull of the vectors (0,1) and (1,0), and so this new commit would follow that prescription (while the current solution, which effectively returns (1,1), doesn't). Also, since all of the tests pass with this commit, we don't have any documented reason to prefer the current convention to the new convention

To summarize, because @yukoba's commit arguably follows numpy's semantics better, returns an element of the subgradient, presumably fixes his real use case, and doesn't break anything I know of, I vote merge!

Is there any use case we're missing?

There might be some other rough edges here, like some inconsistent behavior with lists:

```python
In [3]: grad(lambda x: np.max([x,x]))(1.)
Out[3]: 1.0

In [4]: grad(lambda x: np.max(np.array([x,x])))(1.)  # same as using np.stack
Out[4]: 2.0
```

**Comment by yukoba** [comment]
> The same problem also exists for np.maximum(x, x)

```python
def test_maximum_bug():
    def fun(x):
        return np.maximum([x], [x])
    mat = np.array([1.])
    check_grads(fun, mat)
```

This does not work too.
maximum, minimum, fmax and fmin have the same problem too.

**Comment by dougalm** [comment]
Hi Yukoba, thanks for your PR and for bringing these bugs to our attention! I pushed a couple commits that deal with the problem a slightly different way - spreading the gradient equally between each of the indices whose elements are at the rails, rather than arbitrarily picking one. In Matt's example above, this means taking the gradient to be (0.5, 0.5) rather than (1, 0) or (0, 1). It's still an element of the subgradient, but it's nicely symmetrical. It's also simpler to implement, which is my main reason for preferring it. Does that fix things for you?

**Comment by duvenaud** [comment]
Ah, so satisfying that we don't need a `for` loop!

**Comment by yukoba** [comment]
Wow! Wow! Wow! Wow! Wow! ✌️✌️✌️👍👍👍

Should I modifiy https://github.com/HIPS/autograd/pull/140 ?
